### PR TITLE
Fix usage of outer type argument in extension inline unapply

### LINKED ISSUE #15188: Use of outer type argument in extension inline unapply(Seq) method causes compiler error
## Compiler version

#15053

## Minimized code

```Scala 
object O

extension [T] (ctx: O.type) inline def unapplySeq(input: T): Option[Seq[T]] = Some(Seq(input)) 

@main
def Main = {
  val O(x) = 3
  println(s"x: $x")
}
```

## Output

Compiler error
```
Exception in thread "main" java.lang.IndexOutOfBoundsException: 0
        at scala.collection.LinearSeqOps.apply(LinearSeq.scala:117)
        at scala.collection.LinearSeqOps.apply$(LinearSeq.scala:114)
        at scala.collection.immutable.List.apply(List.scala:79)
        at dotty.tools.dotc.core.Substituters$.substParams(Substituters.scala:151)
        at dotty.tools.dotc.core.Substituters$SubstParamsMap.apply(Substituters.scala:197)
        at dotty.tools.dotc.core.Substituters$SubstParamsMap.apply(Substituters.scala:197)
        at scala.collection.immutable.List.mapConserve(List.scala:472)
        at dotty.tools.dotc.core.Types$TypeMap.mapOverLambda(Types.scala:5553)
        at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:5582)
        at dotty.tools.dotc.core.Substituters$.substParams(Substituters.scala:161)
        at dotty.tools.dotc.core.Substituters$SubstParamsMap.apply(Substituters.scala:197)
        at dotty.tools.dotc.core.Types$TypeMap.mapOverLambda(Types.scala:5555)
        at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:5582)
        at dotty.tools.dotc.core.Substituters$.substParams(Substituters.scala:161)
        ...
```

## Expectation

```
x: 3
```

### PR DESCRIPTION
**Pull request by gorilskij**
Fixes #15191

### CODE REVIEW

**Comment by gorilskij** [comment]
Fixes #15191

**Comment by prolativ** [comment]
@gorilskij I'm a bit confused. Is this a duplicate of https://github.com/lampepfl/dotty/pull/15191 fixing https://github.com/lampepfl/dotty/issues/15188 ?

**Comment by prolativ** [comment]
Somehow I managed to reproduce https://github.com/lampepfl/dotty/issues/15188 on the main branch even without https://github.com/lampepfl/dotty/pull/15191.
But I don't think it would be a good idea to merge a PR knowing that it introduces a bug, especially that there's already a fix for it. Why don't you just merge these 2 PRs then? As the tests are actually the same and the implementations are quite concise I think that would be the most elegant solution in this case.

File: tests/pos/i8577/Main_2.scala [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    val macroC"$xC" = 3
+  }
+
+  {
+    extension (ctx: StringContext) def macroD: MacroD.StringContext = MacroD(ctx)
+    extension [T] (inline ctx: MacroD.StringContext) inline def unapplySeq[U](inline input: T): Option[Seq[T]] =
+      ${ implUnapplyD('ctx, 'input) }
+
+    // miscompilation
+//    val macroD"$xD" = 4
```
Comment (prolativ) [inline tests/pos/i8577/Main_2.scala:34]: If the intentional that this should not compile?
If so, you could create a separate netagive test in `tests/neg/`

File: tests/pos/i8577/Main_2.scala [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    extension (ctx: StringContext) def macroG: MacroG.StringContext = MacroG(ctx)
+    extension [T] (inline ctx: MacroG.StringContext) inline def unapplySeq[U](inline input: T | U): Option[Seq[T | U]] =
+      ${ implUnapplyG('ctx, 'input) }
+
+    // compiler error
+//    val macroG"$xG" = 8
+  }
+}
+
+//  {
```
Comment (prolativ) [inline tests/pos/i8577/Main_2.scala:66]: Why is this section commented out?

**Comment by gorilskij** [comment]
@prolativ you're right, my bad, I thought initially it was dependent but it isn't. Even the code I posted in the issue has the same error with #15191 as without it. These are separate issues, only the test cases of #15189 depend on #15191. They probably should be merged, is there an easy way to do this or should I make a new PR for both?

**Comment by prolativ** [comment]
IMO you can just get all the code changes into one of the PRs and close the other

**Comment by gorilskij** [comment]
Folded into #15191

**Pull request closed without merging**